src/lockup.rs

```diff
@@ -39,18 +39,31 @@ impl Lockup {
         }
     }
 
-    pub fn claim(&mut self, index: LockupIndex) -> LockupClaim {
+    pub fn claim_balance(&mut self, index: LockupIndex, balance_to_claim: Balance) -> LockupClaim {
         let unlocked_balance = self.schedule.unlocked_balance(current_timestamp_sec());
-        assert!(unlocked_balance >= self.claimed_balance, "Invariant");
-        let unclaimed_balance = unlocked_balance - self.claimed_balance;
-        self.claimed_balance = unlocked_balance;
+        let balance_claimed_new = self.claimed_balance + balance_to_claim;
+        assert!(
+            unlocked_balance >= balance_claimed_new,
+            "too big balance_to_claim for lockup {}",
+            index,
+        );
+
+        self.claimed_balance = balance_claimed_new;
         LockupClaim {
             index,
-            unclaimed_balance: unclaimed_balance.into(),
-            is_final: unlocked_balance == self.schedule.total_balance(),
+            balance_to_claim: balance_to_claim.into(),
+            is_final: balance_claimed_new == self.schedule.total_balance(),
         }
     }
 
+    pub fn claim(&mut self, index: LockupIndex) -> LockupClaim {
```

Comment:
I don't see any use of this method, so I suggest to remove it and rename the current `claim_balance` to `claim`